File: packages/python/goldenmatch/goldenmatch/core/autoconfig_memory.py
"""Cross-dataset memory for the auto-config controller.

SQLite-backed store of past committed configs keyed by data-shape signature.
``_initial_config`` consults the memory to short-circuit the v0 heuristic
when a previous run with the same shape converged successfully.

Tier 4 of the zero-config controller spec.
"""
from __future__ import annotations

import hashlib
import logging
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import polars as pl
    from goldenmatch.config.schemas import GoldenMatchConfig

logger = logging.getLogger(__name__)

_SCHEMA = """
CREATE TABLE IF NOT EXISTS autoconfig_runs (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    profile_signature TEXT NOT NULL,
    committed_config_json TEXT NOT NULL,
    succeeded INTEGER NOT NULL,
    n_iterations INTEGER NOT NULL,
    f1_proxy REAL,
    created_at TEXT NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_autoconfig_sig ON autoconfig_runs (profile_signature, created_at DESC);
"""


def profile_signature(df: "pl.DataFrame", *, mode: str = "dedupe") -> str:
    """Compute a coarse shape signature for a DataFrame.

    Same n_cols + same dtype distribution → same signature.
    DBLP-ACM (5 Utf8 cols) ≠ Febrl3 (11 Utf8 cols) ≠ NCVR (10 Utf8 cols).

    Args:
        df: Input DataFrame (target frame in match mode).
        mode: "dedupe" or "match" — dedupe vs match runs are distinct keys even
              with the same shape.

    Returns:
        16-character hex string (truncated SHA-256).
    """
    user_cols = [c for c in df.columns if not c.startswith("__")]
    types = tuple(sorted(str(df.schema[c]) for c in user_cols))
    key = (mode, len(user_cols), types)
    return hashlib.sha256(repr(key).encode()).hexdigest()[:16]


class AutoConfigMemory:
    """SQLite-backed memory of past auto-config runs keyed by data-shape signature.

    Default path: ``~/.goldenmatch/autoconfig_memory.db``
    Tests: pass ``db_path=":memory:"`` for isolated in-memory databases.

    Thread safety: each ``AutoConfigMemory`` instance holds its own
    ``sqlite3.Connection``; do not share instances across threads.
    """

    def __init__(self, db_path: "str | Path | None" = None) -> None:
        if db_path is None:
            default_dir = Path.home() / ".goldenmatch"
            default_dir.mkdir(parents=True, exist_ok=True)
            db_path = default_dir / "autoconfig_memory.db"

        self._db_path = str(db_path)
        self._conn = sqlite3.connect(self._db_path)
        self._conn.executescript(_SCHEMA)
        self._conn.commit()

    # ------------------------------------------------------------------ write

    def remember(
        self,
        signature: str,
        config: "GoldenMatchConfig",
        *,
        succeeded: bool,
        n_iterations: int,
        f1_proxy: float | None = None,
    ) -> None:
        """Persist one completed controller run.

        Args:
            signature: Profile signature from ``profile_signature()``.
            config: The committed ``GoldenMatchConfig`` to store.
            succeeded: True when final health != RED.
            n_iterations: Number of controller iterations completed.
            f1_proxy: Optional proxy metric (mass_above_threshold * (1 - mass_in_borderline)).
        """
        config_json = config.model_dump_json()
        created_at = datetime.now(timezone.utc).isoformat()
        try:
            self._conn.execute(
                """
                INSERT INTO autoconfig_runs
                    (profile_signature, committed_config_json, succeeded,
                     n_iterations, f1_proxy, created_at)
                VALUES (?, ?, ?, ?, ?, ?)
                """,
                (signature, config_json, int(succeeded), n_iterations, f1_proxy, created_at),
            )
            self._conn.commit()
        except sqlite3.Error as exc:
            logger.warning("autoconfig_memory: failed to persist run: %s", exc)

    # ------------------------------------------------------------------ read

    def lookup_best(self, signature: str) -> "GoldenMatchConfig | None":
        """Return the most recent succeeded run's config for this signature.

        Returns None if no successful run exists for this signature.
        """
        from goldenmatch.config.schemas import GoldenMatchConfig

        row = self._conn.execute(
            """
            SELECT committed_config_json
            FROM autoconfig_runs
            WHERE profile_signature = ? AND succeeded = 1
            ORDER BY created_at DESC
            LIMIT 1
            """,
            (signature,),
        ).fetchone()
        if row is None:
            return None
        try:
            return GoldenMatchConfig.model_validate_json(row[0])
        except Exception as exc:
            logger.warning(
                "autoconfig_memory: failed to deserialize cached config for %s: %s",
                signature, exc,
            )
            return None

    def all_for(self, signature: str) -> list[dict]:
        """All runs for a signature, sorted by created_at desc. For diagnostics."""
        rows = self._conn.execute(
            """
            SELECT profile_signature, succeeded, n_iterations, f1_proxy, created_at
            FROM autoconfig_runs
            WHERE profile_signature = ?
            ORDER BY created_at DESC
            """,
            (signature,),
        ).fetchall()
        return [
            {
                "profile_signature": r[0],
                "succeeded": bool(r[1]),
                "n_iterations": r[2],
                "f1_proxy": r[3],
                "created_at": r[4],
            }
            for r in rows
        ]

    # ------------------------------------------------------------------ admin

    def clear(self) -> None:
        """Remove all rows. Primarily for tests."""
        self._conn.execute("DELETE FROM autoconfig_runs")
        self._conn.commit()

    def close(self) -> None:
        """Close the underlying database connection."""
        try:
            self._conn.close()
        except sqlite3.Error:
            pass

File: packages/typescript/goldenmatch/src/core/autoconfigMemory.ts
/**
 * autoconfigMemory.ts — cross-run memory for the auto-config controller.
 * Edge-safe: no `node:` imports (in-memory map; the SQLite persistence in
 * Python's autoconfig_memory.py is a Node concern, deferrable to src/node/).
 *
 * Ports goldenmatch/core/autoconfig_memory.py: a per-shape signature keying a
 * store of past committed configs so the controller can short-circuit the v0
 * heuristic when a previous run with the same shape converged.
 *
 * Signature note: Python hashes `repr((mode, tuple(sorted((name, dtype)))))`.
 * The edge-safe TS port has no polars dtypes, so it derives the per-column
 * type from row values via the same lightweight classifier the rest of the TS
 * core uses. The signature is therefore stable within TS but NOT byte-identical
 * to Python's repr-based hash — the memory short-circuit is a single-language
 * optimization, not a cross-language wire contract.
 */

import type { GoldenMatchConfig, Row } from "./types.js";
import { sha256_16 } from "./memory/hash.js";

/** Lightweight per-column type inference (mirrors the spirit of polars dtype
 *  bucketing for the signature: int / float / bool / str / null). */
function inferColumnType(rows: readonly Row[], col: string): string {
  let sawFloat = false;
  let sawInt = false;
  let sawBool = false;
  let sawStr = false;
  let sawNonNull = false;
  for (const r of rows) {
    const v = (r as Record<string, unknown>)[col];
    if (v === null || v === undefined) continue;
    sawNonNull = true;
    if (typeof v === "boolean") sawBool = true;
    else if (typeof v === "number") {
      if (Number.isInteger(v)) sawInt = true;
      else sawFloat = true;
    } else sawStr = true;
  }
  if (!sawNonNull) return "null";
  if (sawStr) return "str";
  if (sawBool && !sawInt && !sawFloat) return "bool";
  if (sawFloat) return "float";
  if (sawInt) return "int";
  return "str";
}

/**
 * Compute a per-(column-name, dtype) signature for a set of rows. Two row sets
 * hash to the same signature only when they share the same user-facing column
 * names AND inferred per-column types. Mirrors Python ``profile_signature``.
 */
export async function profileSignature(
  rows: readonly Row[],
  mode: "dedupe" | "match" = "dedupe",
): Promise<string> {
  const first = (rows[0] ?? {}) as Record<string, unknown>;
  const cols = Object.keys(first).filter((c) => !c.startsWith("__"));
  const pairs = cols
    .map((c) => [c, inferColumnType(rows, c)] as const)
    .sort((a, b) => a[0].localeCompare(b[0]));
  const key = JSON.stringify([mode, pairs]);
  return sha256_16(key);
}

interface MemoryRow {
  readonly signature: string;
  readonly config: GoldenMatchConfig;
  readonly succeeded: boolean;
  readonly nIterations: number;
  readonly f1Proxy: number | null;
  readonly createdAt: number;
}

/**
 * In-memory store of past auto-config runs keyed by data-shape signature.
 * Mirrors Python ``AutoConfigMemory`` minus SQLite persistence (a Node
 * concern). ``lookupBest`` returns the most recent succeeded config for a
 * signature.
 */
export class AutoConfigMemory {
  private readonly rows: MemoryRow[] = [];
  private seq = 0;

  remember(
    signature: string,
    config: GoldenMatchConfig,
    opts: { succeeded: boolean; nIterations: number; f1Proxy?: number | null },
  ): void {
    this.rows.push({
      signature,
      config,
      succeeded: opts.succeeded,
      nIterations: opts.nIterations,
      f1Proxy: opts.f1Proxy ?? null,
      // Monotonic insert order doubles as created_at for "most recent".
      createdAt: this.seq++,
    });
  }

  lookupBest(signature: string): GoldenMatchConfig | null {
    let best: MemoryRow | null = null;
    for (const r of this.rows) {
      if (r.signature !== signature || !r.succeeded) continue;
      if (best === null || r.createdAt > best.createdAt) best = r;
    }
    return best === null ? null : best.config;
  }

  allFor(signature: string): ReadonlyArray<{
    signature: string;
    succeeded: boolean;
    nIterations: number;
    f1Proxy: number | null;
  }> {
    return this.rows
      .filter((r) => r.signature === signature)
      .sort((a, b) => b.createdAt - a.createdAt)
      .map((r) => ({
        signature: r.signature,
        succeeded: r.succeeded,
        nIterations: r.nIterations,
        f1Proxy: r.f1Proxy,
      }));
  }

  clear(): void {
    this.rows.length = 0;
    this.seq = 0;
  }
}
